Cython/Compiler/TypeSlots.py

```diff
@@ -717,7 +717,7 @@ class DictOffsetSlot(SlotDescriptor):
     def slot_code(self, scope):
         dict_entry = scope.lookup_here("__dict__") if not scope.is_closure_class_scope else None
         if dict_entry and dict_entry.is_variable:
-            if getattr(dict_entry.type, 'cname', None) != 'PyDict_Type':
+            if getattr(dict_entry.type, 'typeptr_cname', None) not in ['&PyDict_Type', '(&PyDict_Type)']:
                 error(dict_entry.pos, "__dict__ slot must be of type 'dict'")
                 return "0"
```

Comment:
Not the fault of this PR, but it seems very late do detect this case here, during code generation.
But shouldn't we just check `is Builtin.dict_type` here? Why look at the type pointer?